src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Indentation/IIndentationService.cs

```diff
@@ -73,7 +47,11 @@ internal static class IIndentationServiceExtensions
         /// </summary>
         public static string GetPreferredIndentation(this SyntaxToken token, Document document, CancellationToken cancellationToken)
         {
+#if CODE_STYLE
+            var sourceText = document.GetTextAsync(cancellationToken).WaitAndGetResult_CanCallOnBackground(cancellationToken);
+#else
             var sourceText = document.GetTextSynchronously(cancellationToken);
```

Comment:
Ah, I hate that the synchronous API is non-public and we end up writing sync helper methods in Workspaces layer that should really be async, and now that we want to move it down you are either left with cascading async flowing up the call chain OR this ugly code with preprocessor directives and `WaitAndGetResult_CanCallOnBackground` calls and such. If possible, try to convert `GetPreferredIndentation` to an async method to avoid this and just use `GetTextAsync` here (if it doesn't lead to too big an explosion of changes).

Reply:
so the main problem here is that for VS we *do* want the actual indentation-feature (e.g. when you hit enter) to be synchronous.  so we want to actually avoid async+blocking.  We then also expose this functionality out for other features to use (like code-fixes).  Codefixes could be async without issue, but we still need the hardened sync path for the mainline feature, and that mainline feature should not block on asynchrony as that could lead to stalls on UI thread. :(

(this is not pretty).